stable_baselines3/common/torch_layers.py

```diff
@@ -224,6 +228,89 @@ def forward(self, features: th.Tensor) -> Tuple[th.Tensor, th.Tensor]:
+    the output features are concatenated and fed through additional MLP network ("combined").
+
+    :param observation_space:
+    :param features_dim: Number of features extracted.
+        This corresponds to the number of unit for the last layer.
+    :param cnn_output_dim: Number of features to output from each CNN submodule(s)
+    :param mlp_output_dim: Number of features to output from each MLP submodule(s)
+    :param mlp_net_arch: Architecture of each MLP network module
+    :param activation_fn: The activation function used in all MLP submodules and combined network
+    :param combined_net_arch: Architecture of the combined network module which calculates the final feature extracted
+    """
+
+    def __init__(
+        self,
+        observation_space: gym.spaces.Dict,
+        features_dim: int = 64,
+        cnn_output_dim: int = 64,
+        mlp_output_dim: int = 64,
+        mlp_net_arch: List[int] = [64, 64],
+        activation_fn: Type[nn.Module] = nn.ReLU,
+        combined_net_arch: List[int] = [64, 64],
+    ):
+        super(CombinedExtractor, self).__init__(observation_space, features_dim=features_dim)
+
+        extractors = {}
+
+        total_concat_size = 0
+        for key, subspace in observation_space.spaces.items():
+            if is_image_space(subspace):
+                # The observation key is an image: create a CNN for it
+                n_input_channels = subspace.shape[0]
+
+                # Nature CNN
+                cnn = nn.Sequential(
+                    nn.Conv2d(n_input_channels, 32, kernel_size=8, stride=4, padding=0),
+                    nn.ReLU(),
+                    nn.Conv2d(32, 64, kernel_size=4, stride=2, padding=0),
+                    nn.ReLU(),
+                    nn.Conv2d(64, 64, kernel_size=3, stride=1, padding=0),
+                    nn.ReLU(),
+                    nn.Flatten(),
+                )
+
+                # TODO is this the best practice for finding out the size?
+                with th.no_grad():
+                    n_flatten = cnn(th.as_tensor(subspace.sample()[None]).float()).shape[1]
+
+                cnn_linear = nn.Sequential(nn.Linear(n_flatten, cnn_output_dim), nn.ReLU())
+
+                extractors[key] = nn.Sequential(*(list(cnn) + list(cnn_linear)))
+
+                total_concat_size += cnn_output_dim
+
+            else:
+                # The observation key is a vector, create a MLP for it
+                extractors[key] = nn.Sequential(
+                    *create_mlp(
+                        subspace.shape[0],
+                        mlp_output_dim,
+                        mlp_net_arch,
+                        activation_fn,
+                        squash_output=False,
+                    )
+                )
+
+                total_concat_size += mlp_output_dim
+
+        self.extractors = nn.ModuleDict(extractors)
+
+        self.combined = nn.Sequential(
```

Comment:
why do you have that?
it looks like a duplicate with the network that will be used after... (and cnn features are already smaller)

Reply:
I wasn't sure what the best approach was. I can remove the combined mlp altogether and return the `th.cat` of the encoded tensors. What would you recommend?

Reply:
yes please, the current structure of SB3 networks are composed of two parts:
the feature extractor (a flatten layer (for vector obs), a CNN or here a combination) and then a mlp used to predict the action, the value from the extracted features.

Reply:
writing that, I realized that we now have an inconsistency with how SB3 works with vector obs (here we have an additional MLP instead of just a flatten layer)